### ISSUE TITLE
Speculate future methods that will need to be jitted via IL analysis of currently jitting methods

### ISSUE THREAD

**Issue by noahfalk** [dotnet/coreclr#19428]
Multicore Jit has been able to provide 20+% startup wins after it is trained, but in many scenarios generating and persisting the training data is problematic. We could instead analyze the IL of methods as they were being generated to infer the additional methods they are likely to call in the near future. As a simple case if we were jitting a method:

```
Foo()
{
    Bar();
    Baz();
}
```

We have a pretty strong expectation that Bar() and Baz() will be needed in the near future. Lets naively assume that each of Foo, Bar, and Baz take 500us to compile, Foo takes 1us to run once jitted, and at the 100us mark of compiling Foo the JIT will call back to the runtime alerting it that Bar and Baz are good speculative compilation candidates. In the current runtime Foo would complete after 1501us whereas in a perfectly parallelized world with no other overhead it would complete in 601us. In practice the wins would probably be less than what we see from trained Multicore Jit, but we'd need further investigation to determine how much less. We might still be able to capture a significant fraction. The advantage is that it doesn't create any dependencies that the application needs to account for which gives it broader scenario reach.

**Comment by AndyAyersMS** [comment]
We would probably need to leverage the logic in multicore jit that prevents / suppresses exceptions from methods that get jitted before they get called, right? Also might there be issues with cctor ordering?

**Comment by noahfalk** [comment]
Yeah I think its definitely something we'd need to investigate if we pursued this. I don't know where we would ultimately want to draw the line in terms of apps that took subtle dependencies on runtime implementation details vs. maximizing performance gains - probably a question we will need some experimentation and feedback to answer properly.

In terms of issues I'd guess the risks look similar to increasing the aggressiveness of the inliner. Speculating that Bar() and Baz() should be compiled before Foo() runs seems roughly the same as inlining Bar + Baz into Foo from the perspective of when we will need to load assemblies, resolve types and run cctors. You may already have some relevant feedback on what that looks like or know why the issues wouldn't be comparable?
